syntax = "proto2";

package voting;

import "google/protobuf/timestamp.proto";

service eVoting {
    rpc PreAuth (VoterName) returns (Challenge);
    rpc Auth (AuthRequest) returns (AuthToken);
    rpc CreateElection (Election) returns (ElectionStatus);
    rpc CastVote (Vote) returns (VoteStatus);
    rpc GetResult(ElectionName) returns (ElectionResult);
}

message Voter {
    required string name = 1;
    required string group = 2;
    required bytes public_key = 3;
}
message VoterName {
    required string name = 1;
}
message VoteStatus {
    required int32 code = 1;
}
message Challenge {
    required bytes value = 1;
}
message Response {
    required bytes value = 1;
}
message AuthToken {
    required bytes value = 1;
}
message AuthRequest {
    required VoterName name = 1;
    required Response response = 2;
}
message Election {
    required string name = 1;
    repeated string groups = 2;
    repeated string choices = 3;
    required google.protobuf.Timestamp end_date = 4;
    required AuthToken token = 5;
}
message Vote {
    required string election_name = 1;
    required string choice_name = 2;
    required AuthToken token = 3;
}
message ElectionName {
    required string name = 1;
}
message VoteCount {
    required string choice_name = 1;
    required int32 count = 2;
    required AuthToken token = 3; 
}
message ElectionResult {
    required int32 status = 1;
    repeated VoteCount count = 2; 
}
message ElectionStatus {
    required int32 code = 1;
}